### ISSUE TITLE
Dropping support for less common platforms

### ISSUE THREAD

**Issue by dsuch** [dsuch/pymqi#166]
Hello,

I would like to ask for your opinion about dropping support for systems other than the ones below:

* Linux
* Mac
* Windows
* z/OS (client connections only)
* AIX

PyMQI has been working on practically every MQ platform in existence for 15+ years but an idea, a good one, that is raised from time to time, is about rewriting parts of PyMQI in either Cython or ctypes rather than using raw C. I prefer Cython.

This would be good because it would make development much, much easier at the cost of uncertainty with regards to systems such as HP-UX.

Myself, I have not seen HP-UX, Tru64 or the like for many, many years and I do not feel much of a need to maintain code for them.

* Using Cython on Linux, Mac and Windows is surely possible.
* z/OS wold be for client connections only so there should not be much trouble
* I am not certain about AIX but I do not have access to the system and neither IBM nor any other user seem interested in granting it

Basically, I am wondering if I have not forgotten about anything in particular? Are there other major systems that need to be included?

I am not going to start such a rewrite anytime soon but I am simply interested in your feedback.

Thanks.

**Comment by chughts** [comment]
If you are interested in joining the MQ Beta Programme, then the Beta Programme Manager could give you access to MQ on AIX as well as MQ on Z/OS.

**Comment by Skyler-Altol** [comment]
I work for an organisation that is potentially one of the world's leading authorities on how to not deal with technical debt (which is as frustrating as it sounds lol), and even we don't have systems of these types. We do have some AIX stuff, but we're moving away from that. 

Personally I'd fully support this, particularly if it makes it easier for you guys to continue building new features, such as the OOP API mentioned in #173

**Comment by kadler** [comment]
FYI, both Cython and ctypes work on AIX, though you may need to manually define and use RTLD_MEMBER if you're loading a library with from an archive. We use it on IBM i for various packages (IBM i has an AIX compatibility layer called PASE).

**Comment by Skyler-Altol** [comment]
I'm also going to throw this out there. Python 2 was EOL at the end of 2019 (1st Jan 2020). We should seriously consider dropping support for Python 2 entirely. This eliminates a *huge* amount of legacy bloat that's in this library, even down in the pymqe module where there's python 2 checks almost every other line. 

I'm well aware that there are some organisations that probably haven't moved, and their stuff will break etc etc, but it's not really fair on you guys to have to continue to offer support for a platform that is as of over a year ago "formally dead". Additionally, Python 3.4 / 3.5 are also both EOL:
https://devguide.python.org/#status-of-python-branches

Therefore we could start putting Python 3.6+ type hints into the library (which far more IDEs support). I won't lie, in the last few days I have been mulling creating a branch in my fork where I just start ripping the guts of this library apart and building against 3.6 as a minimum. RHEL 7 has a max of Python 3.6 from the EPEL repository (CentOS is the same), although Red Hat Software Collections allows for Python 3.8 if your RHEL 7 server is patched to 7.7 or higher. 

Dropping Python 2 for this would be a massive undertaking, and there's certainly areas where I don't have the underlying expertise (Pub/Sub is a major weakpoint for me at the moment, I have very little exposure to it). This certainly wouldn't be a fast thing and I'll likely drop in and out for days or potentially weeks at a time depending on what's happening in life, but I'm genuinely contemplating kicking this off because it's a hell of a way to learn. Being thrown in the deep end is a recurring theme for me in the past few years, and it seems to have worked out quite well.

**Comment by dsuch** [comment]
Yes, I have no problems with dropping support for Python 2.7 - a clear indication in the documentation that such and such version was the last one to support it will be enough.

We need to support Python 3.6 for sure because this is what Ubuntu 18.04 has. Version 3.6 does not have dataclasses but it would not be difficult to bundle [this backport](https://pypi.org/project/dataclasses/) inside PyMQI.

I am definitely all for using dataclasses for code-completion, type hints and static type checks - I use them in other projects and they are a great feature in Python 3.

**Comment by dsuch** [comment]
Mass-closing old tickets.
